doc/man/Qsmtpd.8
.\"TOPICS "Topics:"
.TH Qsmtpd 8 "April 2016" "Qsmtp Version @QSMTP_VERSION@" "Qsmtpd"
.SH NAME
Qsmtpd \- receive mail via (E)SMTP
.SH SYNOPSIS
.B Qsmtpd
[
.I hostname
.I checkprogram
.I subprogram
]
.SH DESCRIPTION
.B Qsmtpd
behaves more or less like
.B qmail-smtpd(8)
, but adds better spam filtering capabilities, better RfC compliance, IPv6 support and more SMTP extensions
out-of-the-box.

.B Qsmtpd
receives mail messages via the Simple Mail Transfer Protocol (SMTP)
and invokes
.B qmail-queue
to deposit them into the outgoing queue.
.B Qsmtpd
must be supplied several environment variables;
see
.BR tcp-environ(5) .

.B Qsmtpd
accepts messages sent as SMTP or ESMTP. The extensions ENHANCEDSTATUSCODES, PIPELINING, 8BITMIME and SIZE 
are always announced in ESMTP mode. CHUNKING is also announced if it has been enabled at compile time. If 
.IR hostname
, 
.IR checkprogram
and 
.IR subprogram
arguments are present and 
.IR checkprogram
exists the AUTH extension is announced. If a TLS certificate is present the STARTTLS extension is also announced.

.B Qsmtpd
counts the number of Recevied:-lines in the mail header.
It rejects any message with 100 or more
.B Received:
header fields.

.B Qsmtpd
can accept LOGIN, PLAIN, and CRAM-MD5 AUTH types.  They can be limited
using the
.IR authtypes
file. It invokes
.IR checkprogram ,
which reads on file descriptor 3 the username, a 0 byte, the password
or challenge derived from
.IR hostname ,
another 0 byte, a CRAM-MD5 response (if applicable to the AUTH type),
and a final 0 byte.
.I checkprogram
invokes
.I subprogram
upon successful authentication, which should in turn return 0 to
.BR Qsmtpd .
.B Qsmtpd
will reject the authentication attempt if it receives a nonzero return
value from
.I checkprogram
or
.IR subprogram .
.SH TRANSPARENCY
.B Qsmtpd
converts the SMTP newline convention into the UNIX newline convention
by converting CR LF into LF.
It returns a permanent error on bare CRs or LFs as well as lines with
more than 1000 characters. Messages that contain non-ASCII characters are
rejected if the client announced the message to be pure ASCII or if the
connection is not in ESMTP mode.

.SH RECIPIENTS
.B Qsmtpd
tries to figure out if local recipients or sender addresses exist. All addresses
that fail this check will be rejected. These tests only work for virtual domains
(e.g. created by vpopmail), local users will not be checked.

.SH "CONTROL FILES"

These files are used for configuring several aspects of
.B Qsmtpd
. Some of them may also be present in the recipient or recipient domain directory
if the domain is controlled by vpopmail. In this case the more local file overrides
the global files. Only files marked with
.I (user)
can be used in user or domain directories.

.TP 4
.I authtypes
A list of all auth types that should be announced. Empty lines and lines beginning
with # are ignored.

.TP 4
.I badmailfrom
.B (user)
Unacceptable envelope sender addresses.
.B Qsmtpd
will reject every recipient address for a message
if the envelope sender address is listed in
.IR badmailfrom .
A line in
.I badmailfrom
may be of the form
.BR @\fIhost\fR ,
meaning every address at
.IR host .
If an entry does not contain a '@' the domain and all subdomains
are rejected.

.EX
    example.com
.EE

would block emails from *@example.com as well as *@*.example.com.
In contrast if the entry does not contain a '@' and begins with '.'
only subdomains will be blocked:

.EX
    .example.org
.EE

would block emails from *@*.example.org, but *@example.org would pass.

.TP 4
.I badcc
.B (user)
Unacceptable CC addresses.
.B Qsmtpd
will reject mails where a recipient address that matches an expression
in this file is already present. The format of the file is the same as for 
.IR badmailfrom .

.TP 4
.I clientca.pem
A list of Certifying Authority (CA) certificates that are used to verify
the client-presented certificates during a TLS-encrypted session.

.TP 4
.I databytes
Maximum number of bytes allowed in a message,
or 0 for no limit.
Default: 0.
If a message exceeds this limit,
.B Qsmtpd
returns a permanent error code to the client;
in contrast, if
the disk is full or
.B Qsmtpd
hits a resource limit, a temporary error code is retuned.

.I databytes
counts bytes as stored on disk, not as transmitted through the network.
It does not count the
.B Qsmtpd
Received line, the
.B qmail-queue
Received line, or the envelope.

.TP 4
.I localiphost
Replacement host name for local IP addresses.
Default:
.IR me ,
if that is supplied.
.B Qsmtpd
is responsible for recognizing dotted-decimal addresses for the
current host.
When it sees a recipient address of the form
.IR box@[d.d.d.d] ,
where
.I d.d.d.d
is a local IP address,
it replaces
.IR [d.d.d.d]
with
.IR localiphost .
This is done before
.IR rcpthosts .

.TP 4
.I rcpthosts
Allowed RCPT domains.
If
.I rcpthosts
is supplied,
.B Qsmtpd
will reject
any envelope recipient address with a domain not listed in
.IR rcpthosts .

.I rcpthosts
may include wildcards:

.EX
   example.org
   .example.com
.EE

The only allowed envelope recipient address without @ sign is postmaster.

.TP 4
.I dh2048.pem
If this 2048 bit Diffie Hellman group is provided,
.B Qsmtpd
will use it for TLS sessions instead of generaring one on-the-fly.
When OpenSSL 1.1.0 or newer is used this file will be ignored and the builtin
DH group from OpenSSL is used instead.

.TP 4
.I rsa2048.pem
If this 2048 bit RSA key is provided,
.B Qsmtpd
will use it for TLS sessions instead of generaring one on-the-fly.

.TP 4
.I servercert.pem.a.a.a.a:b
SSL certificate specific to the IP 
.I a.a.a.a
and the port
.I b
the server is listening on. See 
.IR servercert.pem .
The address must be in dotted decimal form for
IPv4 addresses and in the default IPv6 format (no leading zeroes,
:: abbreviation) for IPv6 addresses.

.TP 4
.I servercert.pem.a.a.a.a
SSL certificate specific to the IP the server is listening on. See 
.IR servercert.pem .
The address must be in dotted decimal form for
IPv4 addresses and in the default IPv6 format (no leading zeroes,
:: abbreviation) for IPv6 addresses.

.TP 4
.I servercert.pem
SSL certificate to be presented to clients in
TLS-encrypted sessions. Certifying Authority
(CA) and intermediate certificates can be added at the end of the file.
Only loaded if no IP-specific certificate file exists.

.TP 4
.I timeoutsmtpd
Number of seconds
.B Qsmtpd
will wait for each new buffer of data from the remote SMTP client.
Default: 1200.

.TP 4
.I tlsclients
A list of email addresses. When relay rules would reject an incoming message,
.B Qsmtpd
can allow it if the client presents a certificate that can be verified against
the CA list in
.I clientca.pem
and the certificate email address is in
.IR tlsclients .

.TP 4
.I tlsserverciphers
A set of OpenSSL cipher strings. Multiple ciphers contained in a
string should be separated by a colon. If the environment variable
.B TLSCIPHERS
is set to such a string, it takes precedence.

.TP 4
.I filterconf
.B (user)
Configures several binary or integer settings. See filterconf(5).

.TP 4
.I forcesslauth
If file contents is a number and not 0, AUTH extension will only be announced
if STARTTLS has been sent by client before and succeeded.

.TP 4
.I vpopbounce
A .qmail-default file with the same contents as this file will be used as sign
that this recipient does not exist. Normally vadddomain(5)

.TP 4
.I rspf
.B (user)
If a domain does not specify a SPF record check for TXT records for this
.I remote SPF 
lists.

.I MAIL FROM:<foo@example.com>
and a rspf containing the line 

.EX
   rspf.example.org
.EE

would cause a DNS TXT lookup for 
.IR example.com.rspf.example.org .
If this query returns a valid SPF string this will be taken as if example.com had this entry itself.

.TP 4
.I spfstrict
.B (user)
List of domains where every mail not explicitely allowed by SPF will be rejected. This means if the 
domain has a SPF entry ending with "~all" the mail will be blocked.

.TP 4
.I spfignore
.B (user)
List of hosts that are known to forward mail to us. Mails from this host will never be blocked because 
of SPF entries.

.TP 4
.I spffriends
Holds a list of IPv4 addresses where no SPF checks will be done at all, neither there will be a
\fIReceived-SPF:\fR-line. Every address is followed by a netmask length specifying if a whole
network is allowed to relay. Use the
.B addipbl
command to add new entries to this file. Matches in this file will override \fIspfstrict\fR.

.TP 4
.I spffriends6
Works the same as
.IR spffriends ,
but for IPv6 addresses.

.TP 4
.I wildcardns
A list of top level domains and their wildcard NS entries. Format is "TLD_IPv6address", where TLD is
the name of the top level domain (without leading dot) and IPv6address is an IPv6 literal. There are
no spaces allowed anywhere in a line containing an entry, but normal comments (lines beginning with #,
empty lines) are allowed. If a TLD has multiple wildcard entries use multiple lines with the same TLD
name and one entry each.

.TP 4
.I authhide
If this file contains a positive integer number the name and IP address of the sending host will not
be written to the \fIReceived:\fR line in the mail body if the client is authenticated. Use this if
your users want some extra privacy.

.TP 4
.I nomail
.B (user)
Reject all mail to this user with the given message. If the file exists but is empty a general
rejection message will be announced. The message may start with a rejection code like:

.EX
550 5.7.1
.EE

The rejection code must be of the form given in the example. The first digit of both blocks must match, 
all other digits may be of any value. The two blocks must be separated by exactly one space. After the 
second block must be at least one more space.

If the code does not match this requirements or is not found at all the code given in the example will be 
used. The rejection line may be of any length, the message will be folded if necessary. The file may contain 
comments, but only one valid line.

.TP 4
.I relayclients

Holds a list of IPv4 addresses allowed for relaying. Every address is followed by a netmask length specifying
if a whole network is allowed to relay. Use the
.B addipbl
command to add new entries to this file.

.TP 4
.I relayclients6
Works the same as
.IR relayclients ,
but for IPv6 addresses.

.SH RELAYING

By default
.B Qsmtpd
does not allow relaying. Contrary to
.B qmail-smtpd
it will not look at the
.I RELAYCLIENT
environment variable.

Relaying is permitted in one of two cases: the user has authenticated himself using SMTP AUTH or a SSL client
certificate, or the IP address of the client is found in one of the
.I relayclients
or
.I relayclients6
control files (see above).

.SH DEBUGGING
If
.B Qsmtpd
has been with the
.I DEBUG_IO
flag the contents of the SMTP transmissions can be recorded. They will
be sent to the syslog daemon with facility mail and log level debug. The contents of the SMTP DATA phase will
never be logged for privacy reasons.

Logging is not enabled by default. If
.B Qsmtpd
finds an environment variable
.I QSMTPD_DEBUG
with a non-empty
value or a file
.I control/Qsmtpd_debug
is readable on startup it will log. Therefore it will usually not harm to
compile that facility into the program.

.SH "SEE ALSO"
tcp-env(1),
filterconf(5),
tcp-environ(5),
qmail-control(5),
qmail-inject(8),
qmail-queue(8),
qmail-smtpd(8)
.SH AUTHOR
Rolf Eike Beer
.SH WEBSITE
https://opensource.sf-tec.de/Qsmtp/


qsmtpd/qsmtpd.c
/** \file qsmtpd.c
 \brief main part of Qsmtpd

 This file contains the main function and the basic commands of Qsmtpd
 SMTP server.
 */

#include <qsmtpd/qsmtpd.h>

#include <control.h>
#include <diropen.h>
#include <log.h>
#include <mmap.h>
#include <netio.h>
#include <qdns.h>
#include <qmaildir.h>
#include <qsmtpd/antispam.h>
#include <qsmtpd/commands.h>
#include <qsmtpd/qsauth.h>
#include <qsmtpd/qsdata.h>
#include <qsmtpd/starttls.h>
#include <qsmtpd/syntax.h>
#include <qsmtpd/userconf.h>
#include <sstring.h>
#include <tls.h>
#include <version.h>

#include <arpa/inet.h>
#include <assert.h>
#include <errno.h>
#include <fcntl.h>
#include <netinet/in.h>
#include <signal.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <sys/file.h>
#include <sys/mman.h>
#include <sys/queue.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <syslog.h>
#include <unistd.h>

#define _C(c, m, f, s, o) { .name = c, .len = sizeof(c) - 1, .mask = m, .func = f, .state = s, .flags = o }

struct smtpcomm *current_command;

static struct smtpcomm commands[] = {
	_C("NOOP",	0xffff, smtp_noop,      -1, 0), /* 0x0001 */
	_C("QUIT",	0xfffd, smtp_quit,       0, 0), /* 0x0002 */
	_C("RSET",	0xfffd, smtp_rset,     0x1, 0), /* 0x0004 */ /* the status to change to is set in smtp_rset */
	_C("HELO",	0xfffd, smtp_helo,       0, 5), /* 0x0008 */
	_C("EHLO",	0xfffd, smtp_ehlo,       0, 5), /* 0x0010 */
	_C("MAIL FROM:",0x0018, smtp_from,       0, 3), /* 0x0020 */
	_C("RCPT TO:",	0x0060, smtp_rcpt,       0, 1), /* 0x0040 */
	_C("DATA",	0x0040, smtp_data,    0x10, 0), /* 0x0080 */ /* the status to change to is changed in smtp_data */
	_C("STARTTLS",	0x0010, smtp_starttls, 0x1, 0), /* 0x0100 */
	_C("AUTH",	0x0010, smtp_auth,      -1, 5), /* 0x0200 */
	_C("VRFY",	0xffff, smtp_vrfy,      -1, 5), /* 0x0400 */
#ifdef CHUNKING
	_C("BDAT",	0x0840, smtp_bdat,      -1, 5), /* 0x0800 */ /* the status to change to is changed in smtp_bdat */
#endif
	_C("POST",	0xffff, http_post,      -1, 1)  /* 0x1000 */ /* this should stay last */
};

#undef _C

unsigned int rcptcount;			/**< number of recipients in lists including rejected */
int relayclient;			/**< flag if this client is allowed to relay by IP: 0 unchecked, 1 allowed, 2 denied */
char *rcpthosts;			/**< memory mapping of control/rcpthosts */
off_t rcpthsize;			/**< sizeof("control/rcpthosts") */
unsigned long sslauth;			/**< if SMTP AUTH is only allowed after STARTTLS */
unsigned long databytes;		/**< maximum message size */
unsigned int goodrcpt;			/**< number of valid recipients */
struct xmitstat xmitstat;		/**< This contains some flags describing the transmission and it's status. */
const char **globalconf;		/**< contents of the global "filterconf" file (or NULL) */
string heloname;			/**< the fqdn to show in helo */
string msgidhost;			/**< the fqdn to use if a message-id is added */
string liphost;				/**< replacement domain if TO address is <foo@[ip]> */
int socketd = 1;			/**< the descriptor where messages to network are written to */
unsigned long comstate = 0x001;		/**< status of the command state machine, initialized to noop */
int authhide;				/**< hide source of authenticated mail */
int submission_mode;			/**< if we should act as message submission agent */

struct recip *thisrecip;

/**
 * \brief write and log error message if opening config file leads to an error
 *
 * @param fn name of the file that caused the error
 * @see err_control2
 */
int
err_control(const char *fn)
{
	const char *logmsg[] = {"error: unable to open file: \"", fn, "\"\n", NULL};

	log_writen(LOG_ERR, logmsg);
	return netwrite("421 4.3.5 unable to read controls\r\n");
}

/**
 * \brief write and log error message if opening config file leads to an error
 *
 * @param msg additional message to log
 * @param fn name of the file that caused the error
 * @see err_control
 */
int
err_control2(const char *msg, const char *fn)
{
	const char *logmsg[] = {"error: unable to open file: ", msg, fn, "\n", NULL};

	log_writen(LOG_ERR, logmsg);
	return netwrite("421 4.3.5 unable to read controls\r\n");
}

/**
 * log error message and terminate program
 *
 * @param error error code that caused the program termination
 */
void
dieerror(int error)
{
	const char *logmsg[] = { "connection from [", xmitstat.remoteip, NULL, NULL, NULL };

	switch (error) {
	case ETIMEDOUT:
		logmsg[2] = "] timed out";
		break;
	case ECONNRESET:
		logmsg[2] = "] died";
		break;
	default:
		logmsg[2] = "] failed, error: ";
		logmsg[3] = strerror(error);
		break;
	}
	log_writen(LOG_WARNING, logmsg);
	conn_cleanup(error);
}

static int
setup(void)
{
	char *tmp;
	unsigned long tl;
	char **tmpconf;
	int rcpthfd;		/* file descriptor of control/rcpthosts */

#ifdef USESYSLOG
	openlog("Qsmtpd", LOG_PID, LOG_MAIL);
#endif

	/* make sure to have a reasonable default timeout if errors happen */
	timeout = 320;

	if (chdir(AUTOQMAIL)) {
		log_write(LOG_ERR, "cannot chdir to qmail directory");
		return EINVAL;
	}

	controldir_fd = get_dirfd(-1, AUTOQMAIL "/control");
	if (controldir_fd < 0) {
		log_write(LOG_ERR, "cannot get a file descriptor for " AUTOQMAIL "/control");
		return EINVAL;
	}

#ifdef DEBUG_IO
	tmp = getenv("QSMTPD_DEBUG");
	if ((tmp != NULL) && (*tmp != '\0'))
		do_debug_io = 1;
	else
		do_debug_io = (faccessat(controldir_fd, "Qsmtpd_debug", R_OK, 0) == 0);
#endif

	int j = loadoneliner(controldir_fd, "me", &heloname.s, 0);
	if (j < 0)
		return errno;
	heloname.len = j;
	if (domainvalid(heloname.s)) {
		log_write(LOG_ERR, "control/me contains invalid name");
		return EINVAL;
	}

	j = loadoneliner(controldir_fd, "msgidhost", &msgidhost.s, 1);
	if (j < 0) {
		j = dupstr(&msgidhost, heloname.s);
		if (j != 0)
			return errno;
	} else {
		msgidhost.len = j;
		if (domainvalid(msgidhost.s)) {
			log_write(LOG_ERR, "control/msgidhost contains invalid name");
			return EINVAL;
		}
	}

	if (( (j = loadoneliner(controldir_fd, "localiphost", &liphost.s, 1)) < 0) && (errno != ENOENT))
		return errno;
	if (j > 0) {
		liphost.len = j;
		if (domainvalid(liphost.s)) {
			log_write(LOG_ERR, "control/localiphost contains invalid name");
			return EINVAL;
		}
	} else {
		liphost.s = heloname.s;
		liphost.len = heloname.len;
	}

	rcpthosts = mmap_name(controldir_fd, "rcpthosts", &rcpthsize, &rcpthfd);

	if (rcpthosts == NULL) {
		int e = errno;

		switch (e) {
		case ENOENT:
			rcpthsize = 0;
			/* fallthrough */
		case 0:
			assert(rcpthsize == 0);
			/* allow this, this just means that no host is local */
			break;
		default:
			log_write(LOG_ERR, "cannot map control/rcpthosts");
			errno = e;
			return -1;
		}
	}

	if ((rcpthsize > 0) && (rcpthsize < 4)) {
		/* minimum length of domain name: x.yy = 4 bytes */
		log_write(LOG_ERR, "control/rcpthosts too short");
		munmap(rcpthosts, rcpthsize);
		rcpthosts = NULL;
		return 1;
	}

#ifdef IPV4ONLY
	tmp = getenv("TCPLOCALIP");
	if (!tmp || !*tmp) {
		log_write(LOG_ERR, "can't figure out local IP (TCPLOCALIP not set)");
		return 1;
	}
	strncpy(xmitstat.remoteip, "::ffff:", sizeof(xmitstat.remoteip));
	strncat(xmitstat.remoteip + strlen("::ffff:"), tmp, sizeof(xmitstat.remoteip) - strlen("::ffff:") - 1);
	if (inet_pton(AF_INET6, xmitstat.remoteip, &xmitstat.slocalip) <= 0) {
		log_write(LOG_ERR, "can't figure out local IP (parse error)");
		return 1;
	}
	strcpy(xmitstat.localip, tmp);

	tmp = getenv("TCPREMOTEIP");
	if (!tmp || !*tmp) {
		log_write(LOG_ERR, "can't figure out IP of remote host (TCPREMOTEIP not set)");
		return 1;
	}
	xmitstat.remoteip[strlen("::ffff:")] = '\0';
	strncat(xmitstat.remoteip + strlen("::ffff:"), tmp, sizeof(xmitstat.remoteip) - strlen("::ffff:") - 1);
	if (inet_pton(AF_INET6, xmitstat.remoteip, &xmitstat.sremoteip) <= 0) {
		log_write(LOG_ERR, "can't figure out IP of remote host (parse error)");
		return 1;
	}
#else /* IPV4ONLY */
	tmp = getenv("TCP6LOCALIP");
	if (!tmp || !*tmp || (inet_pton(AF_INET6, tmp, &xmitstat.slocalip) <= 0)) {
		log_write(LOG_ERR, "can't figure out local IP");
		return 1;
	}
	if (IN6_IS_ADDR_V4MAPPED(&xmitstat.slocalip)) {
		memcpy(xmitstat.localip, tmp + 7, strlen(tmp + 7));
	} else {
		memcpy(xmitstat.localip, tmp, strlen(tmp));
	}

	tmp = getenv("TCP6REMOTEIP");
	if (!tmp || !*tmp || (inet_pton(AF_INET6, tmp, &xmitstat.sremoteip) <= 0)) {
		log_write(LOG_ERR, "can't figure out IP of remote host");
		return 1;
	}
	memcpy(xmitstat.remoteip, tmp, strlen(tmp));
#endif /* IPV4ONLY */

	/* RfC 2821, section 4.5.3.2: "Timeouts"
	 * An SMTP server SHOULD have a timeout of at least 5 minutes while it
	 * is awaiting the next command from the sender. */
	if ( (j = loadintfd(openat(controldir_fd, "timeoutsmtpd", O_RDONLY | O_CLOEXEC), &tl, 320)) ) {
		int e = errno;
		log_write(LOG_ERR, "parse error in control/timeoutsmtpd");
		return e;
	}
	timeout = tl;
	if ( (j = loadintfd(openat(controldir_fd, "databytes", O_RDONLY | O_CLOEXEC), &databytes, 0)) ) {
		int e = errno;
		log_write(LOG_ERR, "parse error in control/databytes");
		return e;
	}
	if (databytes) {
		maxbytes = databytes;
	} else {
		maxbytes = ((size_t)-1) - 1000;
	}
	if ( (j = loadintfd(openat(controldir_fd, "authhide", O_RDONLY | O_CLOEXEC), &tl, 0)) ) {
		log_write(LOG_ERR, "parse error in control/authhide");
		authhide = 0;
	} else {
		authhide = tl ? 1 : 0;
	}

	if ( (j = loadintfd(openat(controldir_fd, "forcesslauth", O_RDONLY | O_CLOEXEC), &sslauth, 0)) ) {
		int e = errno;
		log_write(LOG_ERR, "parse error in control/forcesslauth");
		return e;
	}

	if ( (j = loadlistfd(openat(controldir_fd, "filterconf", O_RDONLY | O_CLOEXEC), &tmpconf, NULL)) ) {
		if ((errno == ENOENT) || (tmpconf == NULL)) {
			tmpconf = NULL;
		} else {
			log_write(LOG_ERR, "error opening control/filterconf");
			return errno;
		}
	}
	globalconf = (const char **)tmpconf;

	j = userbackend_init();
	if (j != 0)
		return j;

	/* Block SIGPIPE, otherwise write errors can't be handled correctly and remote host
	 * will see a connection drop on error (which is bad and violates RfC) */
	sigset_t mask;
	sigemptyset(&mask);
	sigaddset(&mask, SIGPIPE);
	j = sigprocmask(SIG_BLOCK, &mask, NULL);

	relayclient = 0;

	return j;
}

/** initialize variables related to this connection */
static int
connsetup(void)
{
#ifdef IPV4ONLY
	xmitstat.ipv4conn = 1;
#else /* IPV4ONLY */
	xmitstat.ipv4conn = IN6_IS_ADDR_V4MAPPED(&xmitstat.sremoteip) ? 1 : 0;
#endif /* IPV4ONLY */

	int j = ask_dnsname(&xmitstat.sremoteip, &xmitstat.remotehost.s);
	if (j == DNS_ERROR_LOCAL) {
		log_write(LOG_ERR, "can't look up remote host name");
		return -1;
	} else if (j <= 0) {
		STREMPTY(xmitstat.remotehost);
	} else {
		xmitstat.remotehost.len = strlen(xmitstat.remotehost.s);
	}
	xmitstat.remoteinfo = getenv("TCPREMOTEINFO");
	xmitstat.remoteport = getenv("TCPREMOTEPORT");
	if (!xmitstat.remoteport || !*xmitstat.remoteport) {
		log_write(LOG_ERR, "can't figure out port of remote host (TCPREMOTEPORT not set)");
		xmitstat.remoteport = NULL;
	}

	return 0;
}

/**
 * \brief free all ressources allocated for mail transaction
 */
void
freedata(void)
{
	free(xmitstat.mailfrom.s);
	STREMPTY(xmitstat.mailfrom);
	freeips(xmitstat.frommx);
	xmitstat.frommx = NULL;
	free(xmitstat.tlsclient);
	xmitstat.tlsclient = NULL;
	while (!TAILQ_EMPTY(&head)) {
		struct recip *l = TAILQ_FIRST(&head);

		TAILQ_REMOVE(&head, TAILQ_FIRST(&head), entries);
		free(l->to.s);
		free(l);
	}
	rcptcount = 0;
	goodrcpt = 0;
}

/**
 * @brief fork() but clean up internal structures
 *
 * This will work exactly like fork(). If it returns 0 (i.e. you are the
 * child) it will also clean the memory mappings etc. of the Qsmtpd process
 * that the child doesn't need.
 */
pid_t
fork_clean()
{
	pid_t ret = fork();

	if (ret != 0)
		return ret;

	userbackend_free();
	munmap(rcpthosts, rcpthsize);

#ifdef USESYSLOG
	closelog();
#endif /* USESYSLOG */

	if (xmitstat.ssl) {
		int rfd = SSL_get_rfd(xmitstat.ssl);
		int wfd = SSL_get_wfd(xmitstat.ssl);

		/* the fds need to be removed from SSL first, otherwise
		 * destroying the SSL object will terminate the SSL connection
		 * to the remote host. */
		SSL_set_fd(xmitstat.ssl, -1);
		close(wfd);
		if (rfd != wfd)
			close(rfd);

		ssl_free(xmitstat.ssl);
		xmitstat.ssl = NULL;
	}

	return 0;
}

/**
 * \brief clean up the allocated data and exit the process
 * \param rc desired return code of the process
 */
void
conn_cleanup(const int rc)
{
	freedata();
	userbackend_free();
	free(xmitstat.authname.s);

	free(globalconf);
	free(heloname.s);
	free(msgidhost.s);
	exit(rc);
}

static int
smtp_temperror(void)
{
	return netwrite("451 4.3.5 system config error\r\n") ? errno : EDONE;
}

/**
 * line_valid - check if input line contains only syntactically correct characters
 * @returns: 0 on success, else error code
 */
static int
line_valid()
{
	for (size_t i = 0; i < linein.len; i++) {
		/* linein is signed char, so non-ASCII characters are <0 */
		if (linein.s[i] <= 0)
			return EINVAL;
	}
	return 0;
}

static int flagbogus;

static void __attribute__ ((noreturn))
smtploop(void)
{
	badcmds = 0;

	assert(strcmp(commands[1].name, "QUIT") == 0);
	if (!getenv("BANNER")) {
		const char *msg[] = {"220 ", heloname.s, " " VERSIONSTRING " ESMTP", NULL};

		flagbogus = hasinput(0);
		switch (flagbogus) {
		case EBOGUS:
			/* check if someone talks to us like a HTTP proxy and kill the connection if */
			if (!strncmp("POST / HTTP/1.", linein.s, 14)) {
				const char *logmsg[] = {"dropped connection from [", xmitstat.remoteip,
						"]: client is talking HTTP to me", NULL };
				log_writen(LOG_INFO, logmsg);
				exit(0);
			} else {
				/* this is just a broken SMTP engine */
				wait_for_quit();
			}
		case 0:
			flagbogus = -net_writen(msg);
			if (flagbogus == 0)
				break;
			/* fallthrough */
		default:
			/* There was a communication error. Announce temporary error. */
			(void) net_writen(msg);

			(void) net_read(1);

			/* explicitely catch QUIT here: responding with 450 here is bogus */
			if (!strncasecmp(linein.s, commands[1].name, commands[1].len)) {
				if (!linein.s[commands[1].len])
					smtp_quit();
			}
			netwrite("450 4.5.0 transmission error, please try again\r\n");

			/* a conformant client would catch this error, send quit and try again.
			 * So just wait for quit and reject any further command. */
			wait_for_quit();
		}
	}

/* the state machine */
	while (1) {
/* read the line (but only if there is not already an error condition, in this case handle the error first) */
		if (!flagbogus) {
			flagbogus = net_read(1);

/* sanity checks */
			/* we are not in DATA here so there MUST NOT be a non-ASCII character,
			* '\0' is also bogus */
			if (!flagbogus) {
				flagbogus = line_valid();
			} else
				flagbogus = errno;
		}

/* error handling */
		if (flagbogus) {
			check_max_bad_commands();
			/* set flagbogus again in the switch statement to check if an error
			 * occured during error handling. This is a very bad sign: either
			 * we are very short of resources or the client is really really broken */
			switch (flagbogus) {
			case ENOEXEC:	tarpit();
					log_write(LOG_INFO, "bad SMTP command parameter");
					flagbogus = netwrite("501 5.5.2 unrecognized command parameter\r\n") ? errno : 0;
					break;
			case EINVAL:	tarpit();
					log_write(LOG_INFO, "bad SMTP command syntax");
					flagbogus = netwrite("500 5.5.2 command syntax error\r\n") ? errno : 0;
					break;
			case E2BIG:	tarpit();
					log_write(LOG_INFO, "too long SMTP line");
					flagbogus = netwrite("500-5.5.2 line too long\r\n500-5.5.2 This is usually a bug in your mail client\r\n500 5.5.2 Try to use a different encoding like quoted-printable for this mail.\r\n") ? errno : 0;
					break;
			case ENOMEM:	/* ignore errors for the first 2 messages: if the third
						* one succeeds everything is ok */
					netwrite("452-4.3.0 out of memory\r\n");
					sleep(30);
					netwrite("452-4.3.0 give me some time to recover\r\n");
					sleep(30);
					badcmds = 0;
					log_write(LOG_ERR, "out of memory");
					flagbogus = netwrite("452 4.3.0 please try again later\r\n") ? errno : 0;
					break;
			case EIO:	badcmds = 0;
					log_write(LOG_ERR, "IO error");
					flagbogus = netwrite("451 4.3.0 IO error, please try again later\r\n") ? errno : 0;
					break;
			case EMSGSIZE:	badcmds = 0;
					flagbogus = netwrite("552 4.3.1 Too much mail data\r\n") ? errno : 0;
					break;
			case EPROTO:	flagbogus = netwrite("550 5.7.5 data encryption error\r\n") ? errno : 0;
					break;
			case EROFS:	log_write(LOG_ERR, "HELP! queue filesystem looks read only!");
					badcmds = 0;
					flagbogus = netwrite("452 4.3.5 cannot write to queue\r\n") ? errno : 0;
					break;
			case 1:		tarpit();
					flagbogus = netwrite("503 5.5.1 Bad sequence of commands\r\n") ? errno : 0;
					break;
			case EDONE:	badcmds = 0;	/* fallthrough */
			case EBOGUS:	flagbogus = 0;
					break;
			case ECONNRESET:dieerror(flagbogus);
			default:	log_write(LOG_ERR, "writer error. kick me.");
					log_write(LOG_ERR, strerror(flagbogus));
					badcmds = 0;
					flagbogus = netwrite("500 5.3.0 unknown error\r\n") ? errno : 0;
			}
			/* do not work on the command now: it was either not read or was bogus.
			 * Start again and try to read one new to see if it get's better */
			continue;
		}

/* set flagbogus to catch if client writes crap. Will be overwritten if a good command comes in */
		flagbogus = EINVAL;
/* handle the commands */
		for (unsigned int i = 0; i < sizeof(commands) / sizeof(commands[0]); i++) {
			if (!strncasecmp(linein.s, commands[i].name, commands[i].len)) {
				if (comstate & commands[i].mask) {
					const long ostate = commands[i].state; /* the state originally recorded for this command */

					/* "space required" may not come without "takes arguments" */
					assert((commands[i].flags & 5) != 4);
					if (!(commands[i].flags & 2) && (linein.len > 510)) {
						/* RfC 2821, section 4.5.3.1 defines the maximum length of a command line
						 * to 512 chars if this limit is not raised by an extension. Since we
						 * stripped CRLF our limit is 510. A command may override this check and
						 * check line length by itself */
						 flagbogus = E2BIG;
						 break;
					}

					if (!(commands[i].flags & 1) && linein.s[commands[i].len]) {
						flagbogus = EINVAL;
					} else if ((commands[i].flags & 4) && (linein.s[commands[i].len] != ' ')) {
						flagbogus = EINVAL;
					} else {
						current_command = commands + i;
						flagbogus = commands[i].func();
						current_command = NULL;
					}

					/* command succeded */
					if (!flagbogus) {
						if (commands[i].state > 0)
							comstate = commands[i].state;
						else if (!commands[i].state)
							comstate = (1 << i);
						badcmds = 0;
					}
					commands[i].state = ostate;	/* in case a command has changed that */
				} else
					flagbogus = 1;
				break;
			}
		}
	}
}

int
main(int argc, char **argv)
{
	const char *localport = getenv("TCPLOCALPORT");

	if (setup()) {
		/* setup failed: make sure we wait until the "quit" of the other host but
		 * do not process any mail. Commands RSET, QUIT and NOOP are still allowed.
		 * The state will not change so a client ignoring our error code will get
		 * "bad sequence of commands" and will be kicked if it still doesn't care.
		 * The HTTP POST stub is also kept enabled.
		 */
		for (unsigned int i = 3; i < (sizeof(commands) / sizeof(commands[0])) - 1; i++) {
			commands[i].func = smtp_temperror;
			commands[i].state = -1;
		}
	} else {
		STREMPTY(xmitstat.authname);
		xmitstat.check2822 = 2;
		TAILQ_INIT(&head);		/* Initialize the recipient list. */
	}

	submission_mode = (localport != NULL) && (strcmp(localport, "587") == 0);

	/* Assume all given parameters are for auth checking */
	auth_setup(argc, (const char **) argv);

	if (connsetup() < 0)
		flagbogus = errno;
	smtploop();
}
